### DISCUSSION TITLE
[RFC] Button support `variant` and `color` to deprecate `type` and `danger`

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#49700] [RFCs]
## Motivation

![image](https://github.com/ant-design/ant-design/assets/27722486/365b67e1-a1de-43e0-8c40-18f2118dccc4)

For a long time, Button of antd lacks a secondary variants like this. We had provide `type` as variants for users, but `primary` in `type` makes it impossible to both have `filled` and `primary` variants.

![image](https://github.com/ant-design/ant-design/assets/27722486/7a8001a7-6681-4114-9c75-c7fbf9ff81fc)

## API

We will provide two brand new props to describe Button's variant, which are `variants` and `color`.

| Props | Description | Type | Default |
| --- | --- | --- | --- |
| variant | Variants of Button | `'outlined' \| 'dashed' \| 'solid' \| 'filled' \| 'text' \| 'link'` | `'outlined'` |
| color | Main color of Button | `'primary' \| 'danger' \| 'default'` ( Maybe could include `success` and `warning`) | `'default'` |

## Detail

By combine the two props above, we can get more variants of Button.

![image](https://github.com/ant-design/ant-design/assets/27722486/c93720c4-eb80-4fdb-9395-bee1b9579b77)

## How to teach

We should preserve `type` and `danger` in 5.x, make them warn that would be removed in next major.
The usage in demos should be migrated to `variant` and `color`.

## Drawbacks

We should also take `ghost` into consideration.

**Comment by arvinxx** [comment]
solid 是不是可以考虑这种：

<img width="323" alt="image" src="https://github.com/ant-design/ant-design/assets/28616219/4af4e308-320b-4485-895f-9812418a2039">

**Reply by MadCcc** [reply]
这个我理解也属于 outlined，只是 bg 换了一下

**Comment by arvinxx** [comment]
manitne 的 button （https://mantine.dev/core/button/#usage）， link 对应的是 transparent ，text 对应的是 subtle，然后有一种叫 white 变体缺失。效果如下：

| 默认态| hover态| 
| --- | --- |
| <img width="649" alt="image" src="https://github.com/ant-design/ant-design/assets/28616219/6b6545bf-d0ed-4d18-b859-f6a03b6ab1ca"> | <img width="656" alt="image" src="https://github.com/ant-design/ant-design/assets/28616219/08d4f3dd-9db2-4efb-bf2d-8d2845cb88ae"> |

要看下这个状态有没有必要加。


另外有点感觉 text 和 link 这个名称，和 antd 样式上的效果对应得不是很好。看看有没有必要改个语义化更准确的名字？

（mantine 的变体命名感觉也不是很好）

**Reply by MadCcc** [reply]
white 其实就是把 text 的 hover 背景统一成一个灰色，实际可以用 token 解决，可以不用考虑。

**Comment by arvinxx** [comment]
最后一点，我感觉可能 type 保留也有必要。比如之前只需要 `type='primary'` 就可以做出最重视觉强度的按钮，但新的方案里需要 `variant="solid" color="primary"` 写两个，这个感觉不是很好，毕竟 primary button 还是非常常用的。

**Reply by arvinxx** [reply]
可以考虑把 type 的作用定义为设计系统预先设定了的一组 `variant + color` 的集合。 比如:

- `type='primary'` = `variant:"solid"  + color:"primary"` 
-  `type='danger'` = `variant:"solid"  + color:"danger"`
-  `type='link'` =  `variant:"link"  + color:"primary"`


如果用户传了 type ，那么组件优先使用 type 所定义的变体类型。但如果用户需要自定义，那么将type 改为 color 以后，再结合 variant 使用？

**Reply by MadCcc** [reply]
api 都保留，实际内部实现都换成 `variant` 和 `color` 就可以

**Comment by MadCcc** [comment]
`color` -> `primary` `red` `grey`

**Comment by MadCcc** [comment]
solid & default => 
![image](https://github.com/ant-design/ant-design/assets/27722486/242726ac-d177-4af7-9cbc-5d33f8df7952)

**Comment by crazyair** [comment]
要不要支持个 noPadding

**Comment by coding-ice** [comment]
I want to try this feature.
